### DISCUSSION TITLE
How does processing work with multiple span processors?

### DISCUSSION THREAD

**Discussion by DenisAngeletta** [open-telemetry/opentelemetry-java-instrumentation#6158] [Q&A]
Hi everyone!

Having multiple span processors added to the TracerProvider is possible. I wanted to ask how processing looks like in that case. I'd imagine something like:
`onStart(1) -> onEnd(1) -> onStart(2) -> onEnd(2)`

There are multiple ways to add SpanProcessors as far as I know:

1:
```java
@Override
public void customize(AutoConfigurationCustomizer autoConfiguration) {
    autoConfiguration.addTracerProviderCustomizer(
        (sdkTracerProviderBuilder, configProperties) -> sdkTracerProviderBuilder.addSpanProcessor(new SpanProcessorOne())
                .addSpanProcessor(new SpanProcessorTwo()));
}
```
2:
```java
@Override
public void customize(AutoConfigurationCustomizer autoConfiguration) {
    autoConfiguration.addTracerProviderCustomizer(
        (sdkTracerProviderBuilder, configProperties) -> sdkTracerProviderBuilder.addSpanProcessor(
    SpanProcessor.composite(new SpanProcessorOne(), new SpanProcessorTwo())));
}
```

I'm not sure how (1) would work but (2) essentially creates a MultiSpanProcessor in which to my understanding onStart is called twice before any call of onEnd. Am I understanding this correctly?

The reason I am asking this because I need a workaround where I setAttributes when a Span is ending. If the processing of spans happens like `onStart(1) -> onEnd(1) -> onStart(2) -> onEnd(2)` then I would also be happy.

**Comment by laurit** [comment]
When span is started `onStart` is called on all span processors when span is ended `onEnd` is called. These methods are called synchronously.
I believe that whichever way you add the span processors they'll be wrapped in a `MultiSpanProcessor`.
